api/build/auto_cancel.go

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package build
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-vela/server/database"
+	"github.com/go-vela/server/internal/token"
+	"github.com/go-vela/types/constants"
+	"github.com/go-vela/types/library"
+)
+
+// AutoCancel is a helper function that checks to see if any pending or running
+// builds for the repo can be replaced by the current build.
+func AutoCancel(c *gin.Context, b *library.Build, rBs []*library.Build, r *library.Repo, pending, running bool) error {
+	// iterate through pending and running builds
+	for _, rB := range rBs {
+		// if build is the current build, continue
+		if rB.GetID() == b.GetID() {
+			continue
+		}
+
+		// ensure criteria is met before auto canceling
+		if (strings.EqualFold(rB.GetEvent(), constants.EventPush) &&
+			strings.EqualFold(b.GetEvent(), constants.EventPush) &&
+			strings.EqualFold(b.GetBranch(), rB.GetBranch())) ||
+			(strings.EqualFold(rB.GetEvent(), constants.EventPull) &&
+				strings.EqualFold(b.GetEventAction(), rB.GetEventAction()) &&
+				strings.EqualFold(b.GetHeadRef(), rB.GetHeadRef())) {
```

Comment:
should we add a helper func? or comments? this `if` is gnarly and my eyes glaze over it :(